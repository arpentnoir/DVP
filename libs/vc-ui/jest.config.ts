/* eslint-disable */
export default {
  displayName: 'vc-ui',
  preset: '../../jest.preset.js',
  setupFiles: ['./jest.setup.ts'],
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/vc-ui',
  moduleNameMapper: {
    '^react-pdf': 'react-pdf/dist/umd/entry.jest',
  },
};
