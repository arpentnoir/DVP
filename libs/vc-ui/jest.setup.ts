import 'jest-canvas-mock';
import '@testing-library/jest-dom';

const mockClipboard = {
  writeText: jest.fn(),
};

(global.navigator.clipboard as any) = mockClipboard;
