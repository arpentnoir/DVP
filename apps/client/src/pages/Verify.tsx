import {
  VerifiableCredential,
  WrappedVerifiableCredential,
} from '@dvp/api-interfaces';
import { CertificateUpload, isVerifiableCredential, Text } from '@dvp/vc-ui';
import { Alert, Box, CircularProgress, Stack } from '@mui/material';
import React, { useEffect, useMemo, useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { ROUTES } from '../constants';
import { BaseLayout } from '../layouts';
import { fetchAndDecryptVC } from '../services/storage';
import { verify } from '../services/vc';

const ACCEPT_MESSAGE = 'Only JSON files with max size of 3 MB';
const MAX_FILE_SIZE = 3 * 1000 * 1000; // 3 Megabytes

export const Verify = () => {
  const [error, setError] = useState(false);
  const [loading, setLoading] = useState(false);
  const [uploadErrorMessage, setUploadErrorMessage] = useState('');
  const [errorKeyId, setErrorKeyId] = useState('');
  const location = useLocation();
  const navigate = useNavigate();
  const verifyRef = React.createRef<HTMLParagraphElement>();

  const setErrorMessage = (message: string) => {
    setUploadErrorMessage(message);
    /**
     * Accesibility: the upload error message needs to be read out
     * every time. As such, to force a re-render of the upload component
     * a random number is generated and assigned as the key.
     **/

    setErrorKeyId(Math.random().toString());
  };

  const handleFiles = (files: FileList) => {
    const reader = new FileReader();
    reader.onload = async () => {
      try {
        setLoading(true);

        // Check that file is of JSON type
        const document = JSON.parse(
          reader.result as string
        ) as VerifiableCredential;

        // Check that file is a verifiable credential
        if (!isVerifiableCredential(document as WrappedVerifiableCredential)) {
          setErrorMessage('Must be a Verifiable Credential');
          setLoading(false);
          return;
        }

        const result = await verify(document);
        navigate(ROUTES.VIEWER, {
          state: result,
        });

        setLoading(false);
      } catch (err) {
        if (err instanceof SyntaxError) {
          setErrorMessage('Must be a JSON file');
        } else {
          setError(true);
        }

        setLoading(false);
      }
    };

    if (files[0].size > MAX_FILE_SIZE) {
      setErrorMessage('Max file size exceeded');
      return;
    }

    reader.readAsText(files[0] as Blob);
  };

  useEffect(() => {
    if (location.search !== '') {
      setLoading(true);
      fetchAndDecryptVC(location.search)
        .then(verify)
        .then((result) => {
          navigate(ROUTES.VIEWER, {
            state: result,
          });
        })
        .catch(() => {
          setLoading(false);
          setError(true);
        });
    }
  }, []);

  const displayResult = useMemo(() => {
    return (
      <>
        {error && <ErrorMessage />}
        <CertificateUpload
          handleFiles={handleFiles}
          errorMessage={uploadErrorMessage}
          key={errorKeyId.toString()}
          acceptMessage={ACCEPT_MESSAGE}
        />
      </>
    );
  }, [error, uploadErrorMessage, errorKeyId]);

  return (
    <BaseLayout title="Verify a document">
      <Stack
        sx={{ display: 'flex', height: '100%' }}
        spacing={{ xs: '16px', md: 0 }}
      >
        <Box>
          <Text variant="h4" fontWeight="bold" paddingBottom="24px">
            Verify a document
          </Text>
          <Text>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam eu
            turpis molestie, dictum est a, mattis tellus. Sed dignissim, metus
            nec fringilla accumsan, risus sem sollicitudin lacus, ut interdum
            tellus elit sed risus. Maecenas eget condimentum velit, sit amet
            feugiat lectus. Class aptent taciti sociosqu ad litora torquent per
            conubia nostra, per inceptos himenaeos. Praesent auctor purus luctus
            enim egestas, ac scelerisque ante pulvinar. Donec ut rhoncus ex.
            Suspendisse ac rhoncus nisl, eu tempor urna. Curabitur vel bibendum
            lorem. Morbi convallis convallis diam sit amet lacinia. Aliquam in
            elementum tellus.
          </Text>
        </Box>
        <Box
          sx={{
            flex: 1,
          }}
        >
          <Stack
            height="100%"
            maxWidth="800px"
            justifyContent="center"
            margin="auto"
          >
            {loading ? (
              <Stack
                sx={{
                  height: '100%',
                  alignItems: 'center',
                  justifyContent: 'center',
                }}
              >
                <CircularProgress />
                <Text
                  variant="h5"
                  paddingTop={3}
                  ref={verifyRef}
                  aria-label="Please wait while your document is being verified"
                >
                  Verifying Document <span aria-hidden>...</span>
                </Text>
              </Stack>
            ) : (
              displayResult
            )}
          </Stack>
        </Box>
      </Stack>
    </BaseLayout>
  );
};

const ErrorMessage = () => (
  <Stack alignItems="center" paddingY="16px">
    <Alert severity="error" tabIndex={0}>
      <Stack alignItems="center">
        <Text>There was an error verifying your document</Text>
        <Text variant="body1" fontWeight="bold" paddingTop="12px">
          Please rescan the QR and try again
        </Text>
      </Stack>
    </Alert>
  </Stack>
);
