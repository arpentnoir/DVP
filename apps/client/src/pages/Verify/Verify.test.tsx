import { render } from '@dvp/vc-ui';
import { Verify } from './Verify';

describe('Verify', () => {
  it('should render correctly', () => {
    const { baseElement } = render(<Verify />);

    expect(baseElement).toMatchSnapshot();
  });
});
