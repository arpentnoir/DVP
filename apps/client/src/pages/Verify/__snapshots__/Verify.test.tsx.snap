// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify should render correctly 1`] = `
<body>
  <div>
    <div
      class="MuiContainer-root MuiContainer-disableGutters css-1y9y4eh-MuiContainer-root"
    >
      <div
        class="css-13fkgki-MuiStack-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <h1
            class="MuiTypography-root MuiTypography-h1 css-1ne5dp9-MuiTypography-root"
            tabindex="0"
          >
            Verify a document
          </h1>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            tabindex="0"
          >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam eu turpis molestie, dictum est a, mattis tellus. Sed dignissim, metus nec fringilla accumsan, risus sem sollicitudin lacus, ut interdum tellus elit sed risus. Maecenas eget condimentum velit, sit amet feugiat lectus. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Praesent auctor purus luctus enim egestas, ac scelerisque ante pulvinar. Donec ut rhoncus ex. Suspendisse ac rhoncus nisl, eu tempor urna. Curabitur vel bibendum lorem. Morbi convallis convallis diam sit amet lacinia. Aliquam in elementum tellus.
          </p>
        </div>
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <div
            class="css-yebffi-MuiStack-root"
          >
            <div
              class="css-12a1c8v-MuiStack-root"
            >
              <div
                class="MuiBox-root css-xi606m"
                id="certificate-upload"
              >
                <div
                  class="css-1g3jqqf-MuiStack-root"
                >
                  <div
                    class="css-1wrk7t-MuiStack-root"
                  >
                    <div
                      class="css-1i78psr-MuiStack-root"
                    >
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1eoulhv-MuiButtonBase-root-MuiButton-root"
                        data-testid="button:Scan QR code"
                        tabindex="0"
                        type="button"
                      >
                        <div
                          class="css-doi3ij-MuiStack-root"
                        >
                          <h4
                            class="MuiTypography-root MuiTypography-h4 css-5ukp3i-MuiTypography-root"
                            data-testid="button:Scan QR code:label"
                            tabindex="-1"
                          >
                            Scan QR code
                          </h4>
                        </div>
                        <span
                          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        />
                      </button>
                      <div
                        class="MuiBox-root css-1xu3tth"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 css-1sx1eu0-MuiTypography-root"
                          tabindex="-1"
                        >
                          Or Â 
                        </p>
                        <div
                          aria-controls="filename"
                          aria-label="Select a file. Only JSON files with max size of 3 MB"
                          class="MuiBox-root css-13imy7l"
                          role="button"
                          tabindex="0"
                        >
                          Select a file
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-1g6u8k6-MuiInputBase-root-MuiInput-root"
                >
                  <input
                    class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                    data-testid="input-file-upload"
                    id="input-file-upload"
                    type="file"
                    value=""
                  />
                </div>
                <label
                  aria-label="Upload file"
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-1l91s8u-MuiFormLabel-root-MuiInputLabel-root"
                  for="input-file-upload"
                  id="label-file-upload"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
